python/lsst/ts/hexrotcomm/command_telemetry_client.py

```diff
@@ -419,9 +417,60 @@ async def put_command(self, command):
                 f"command={command!r} must be an instance of structs.Command"
             )
 
-        # Set command.counter to the next value (starting from 1).
-        # Note: this code allows command.counter to wrap around,
-        # without having to know the data type of that field.
-        command.counter = self._command_counts.get(command.code, 0) + 1
-        self._command_counts[command.code] = command.counter
-        await tcpip.write_from(self.command_writer, command)
+        async with self.command_lock:
+            command.counter = self._last_command.counter + 1
+            self._last_command = command
+            await tcpip.write_from(self.command_writer, command)
+
+            self.read_command_status_task = asyncio.create_task(
+                asyncio.wait_for(
+                    self.read_command_status(command.counter),
+                    timeout=COMMAND_STATUS_TIMEOUT,
+                )
+            )
+            await self.read_command_status_task
+            if self.command_status.status == CommandStatusCode.ACK:
+                return self.command_status.duration
+            elif self.command_status.status == CommandStatusCode.NO_ACK:
+                reason = self.command_status.reason.decode()
+                raise salobj.ExpectedError(reason)
+            else:
+                raise salobj.ExpectedError(
+                    f"Unknown command status {self.command_status.status}; "
+                    "low-level command assumed to have failed"
+                )
+
+    async def read_command_status(self, counter):
+        """Wait for a command status message whose counter matches
+        a specified value.
+        """
+        try:
+            while True:
+                if not self.command_connected:
+                    raise RuntimeError("Command socket disconnected during the command")
```

Comment:
Please consider to put the `RuntimeError ` to the doc string.

Reply:
I have updated the code to use the counter in the header.

(I look forward to json where everything is contained in the same message. Though it's not a high priority.)